lib/history.js

```diff
@@ -0,0 +1,111 @@
+'use strict';
+
+var _ = require('lodash');
+var LocalStorage = require('node-localstorage').LocalStorage;
+var localStorage = new LocalStorage('./.cmd_history');
+
+// Number of command histories kept in persistent storage
+var HISTORY_SIZE = 50;
+
+var History = function () {
+  // Prompt Command History
+  // Histctr moves based on number of times 'up' (+= ctr)
+  //  or 'down' (-= ctr) was pressed in traversing
+  // command history.
+  this._hist = [];
+  this._histCtr = 0;
+
+  // When in a 'mode', we reset the
+  // history and store it in a cache until
+  // exiting the 'mode', at which point we
+  // resume the original history.
+  this._histCache = [];
+  this._histCtrCache = 0;
+};
+
+/**
+ * Initialize the history with local storage data
+ * Called when session is initialized
+ */
+History.prototype.init = function () {
+  // Load history from local storage
+  // Read cmd_history_cache first in case we exited from a mode last time.
+  var persistedHistory = JSON.parse(localStorage.getItem('cmd_history'));
+  if (_.isArray(persistedHistory)) {
+    Array.prototype.push.apply(this._hist, persistedHistory);
+  }
+};
+
+/**
+ * Get previous history. Called when up is pressed.
+ * @return {String}
+ */
+History.prototype.getPreviousHistory = function () {
+  this._histCtr++;
+  this._histCtr = (this._histCtr > this._hist.length) ? this._hist.length : this._histCtr;
+  return this._hist[this._hist.length - (this._histCtr)];
+};
+
+/**
+ * Get next history. Called when down is pressed.
+ * @return {String}
+ */
+History.prototype.getNextHistory = function () {
+  this._histCtr--;
+
+  // Return empty prompt if the we dont have any history to show
+  if (this._histCtr < 1) {
+    this._histCtr = 0;
+    return '';
+  }
+
+  return this._hist[this._hist.length - (this._histCtr)];
+};
+
+/**
+ * A new command was submitted. Called when enter is pressed and the prompt is not empty.
+ * @param cmd
+ */
+History.prototype.newCommand = function (cmd) {
+  this._hist.push(cmd);
+  this._histCtr = 0;
+
+  // Only persist history when not in mode
+  if (!this._inMode) {
+    var persistedHistory = this._hist;
+    var historyLen = this._hist.length;
+    if (historyLen > HISTORY_SIZE) {
+      persistedHistory = this._hist.slice(historyLen - HISTORY_SIZE - 1, historyLen - 1);
+    }
+
+    // Add to local storage
+    localStorage.setItem('cmd_history', JSON.stringify(persistedHistory));
```

Comment:
The history needs to be separated out based on application. What if someone was running two Vorpal apps on the same computer? They would share the same command history, pulling up commands not even registered in that instance of Vorpal.

I think we should be able to pass a unique ID into Vorpal's constructor for history:

```
var vorpal = new Vorpal();
vorpal
  .delimiter('myfoobarapp: ')
  .history('dkt48d')
  .show();
```

This would activate persistent history and keep it separated. One would use the UID like this:

```
localStorage.setItem('cmd_history_' + this.historyUID, JSON.stringify(persistedHistory)); 
```